### PR TITLE
[SPARK-17472] [PYSPARK] Better error message for serialization failures of large objects in Python

### PR DESCRIPTION
**Pull request by ericl**
## What changes were proposed in this pull request?

For large objects, pickle does not raise useful error messages. However, we can wrap them to be slightly more user friendly:

Example 1:

```
def run():
  import numpy.random as nr
  b = nr.bytes(8 * 1000000000)
  sc.parallelize(range(1000), 1000).map(lambda x: len(b)).count()

run()
```

Before:

```
error: 'i' format requires -2147483648 <= number <= 2147483647
```

After:

```
pickle.PicklingError: Object too large to serialize: 'i' format requires -2147483648 <= number <= 2147483647
```

Example 2:

```
def run():
  import numpy.random as nr
  b = sc.broadcast(nr.bytes(8 * 1000000000))
  sc.parallelize(range(1000), 1000).map(lambda x: len(b.value)).count()

run()
```

Before:

```
SystemError: error return without exception set
```

After:

```
cPickle.PicklingError: Could not serialize broadcast: SystemError: error return without exception set
```
## How was this patch tested?

Manually tried out these cases

cc @davies

### CODE REVIEW

**Comment by ericl** [comment]
Hmm, seems the linter doesn't like the raise-with-traceback syntax.

**Comment by ericl** [comment]
The issue seems to be Python 3 compatibility. Unfortunately, since the syntax for raising exceptions with traceback has changed across versions, there might not be a straightforward workaround.

**Comment by ericl** [comment]
dropped traceback propagation

File: python/pyspark/broadcast.py [hunk 1]
```diff
@@ -75,7 +75,13 @@ def __init__(self, sc=None, value=None, pickle_registry=None, path=None):
...
 
     def dump(self, value, f):
-        pickle.dump(value, f, 2)
+        try:
+            pickle.dump(value, f, 2)
+        except pickle.PickleError:
+            raise
+        except Exception as e:
+            msg = "Could not serialize broadcast: " + e.__class__.__name__ + ": " + e.message
+            raise pickle.PicklingError(msg)
```
Comment (davies) [inline python/pyspark/broadcast.py]: Can we log the stacktrace here?

File: python/pyspark/cloudpickle.py [hunk 2]
```diff
@@ -109,6 +109,15 @@ def dump(self, obj):
...
                 msg = """Could not pickle object as excessively deep recursion required."""
                 raise pickle.PicklingError(msg)
+        except pickle.PickleError:
+            raise
+        except Exception as e:
+            if "'i' format requires" in e.message:
+                msg = "Object too large to serialize: " + e.message
+            else:
+                msg = "Could not serialize object: " + e.__class__.__name__ + ": " + e.message
+            raise pickle.PicklingError(msg)
```
Comment (davies) [inline python/pyspark/cloudpickle.py]: Same here

**Comment by davies** [comment]
Once we could log the original stacktrace, this looks good to me.

File: python/pyspark/broadcast.py (hunk 1 above)
Comment (Stibbons) [inline python/pyspark/broadcast.py]: it is indeed a good practice to log when wrapping exception is more contextual one. logging.exception(msg) will be perfect

Comment (ericl) [inline python/pyspark/broadcast.py]: Is there a standard logging module in pyspark (didn't find one), or should I just print to stderr?

**Comment by ericl** [comment]
Updated to also print out the original trace to stderr

File: python/pyspark/broadcast.py (hunk 1 above)
Comment (Stibbons) [inline python/pyspark/broadcast.py]: i would hope "logging" would work

Comment (ericl) [inline python/pyspark/broadcast.py]: It seems we use print_exec() elsewhere so going to use that for consistency.

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master, thanks!

**Pull request closed without merging**